src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,324 @@
+  DGLContext ctx,
+  DGLDataType fidtype,
+  const IdxType max_degree,
+  const IdxType num_rows,
+  const int importance_sampling,
+  const bool weights,
+  const IdxType* rows_data,
+  const IdxType* indptr,
+  const FloatType* A,
+  const IdxType* indices,
+  const IdxType num_picks,
+  const FloatType* ds,
+  FloatType* cs,
+  phmap::flat_hash_map<IdxType, FloatType> &hop_map
+) {
+  constexpr FloatType ONE = 1;
+  // ps stands for \pi in arXiv:2210.13339
+  FloatArray ps_array = NDArray::Empty({max_degree + 1}, fidtype, ctx);
+  FloatType* ps = ps_array.Ptr<FloatType>();
+
+  double prev_ex_nodes = max_degree * num_rows;
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+  for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+    /*
+    An implementation detail about when there are no weights, then the first c values can be computed in O(1) because all the edge weights are uniform. The c value is just k / d where k is fanout and d is the degree. But when there are weights, since the computed c values are incorrect, I have to skip some logic making use of it to compute the right c values below.
+
+    The later iterations will have correct c values so they will take the if in that case.
+
+    I also increase importance_sampling by 1 above so that the weighted case can do one more of this iteration to match the unweighted case.
+    */
+    if (!weights || iters) {
+      hop_map2.clear();
+      for (int64_t i = 0; i < num_rows; ++i) {
+        const FloatType c = cs[i];
+        const IdxType rid = rows_data[i];
+        for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+          const auto ct = c * (weights && iters == 1 ? A[j] : 1);
+          auto itb = hop_map2.emplace(indices[j], ct);
+          if (!itb.second)
+            itb.first->second = std::max(ct, itb.first->second);
+        }
+      }
+      if (hop_map.empty())
+        hop_map = std::move(hop_map2);
+      else
+        for (auto it : hop_map2)
+          hop_map[it.first] *= it.second;
+    }
+
+    for (int64_t i = 0; i < num_rows; ++i) {
+      const IdxType rid = rows_data[i];
+      const auto d = indptr[rid + 1] - indptr[rid];
+      if (d == 0)
+        continue;
+
+      const auto k = std::min(num_picks, d);
+
+      if (hop_map.empty()) {  // weights first iter, pi = A
+        for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+          ps[j - indptr[rid]] = A[j];
+      } else {
+        for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+          ps[j - indptr[rid]] = hop_map[indices[j]];
+      }
+
+      // stands for right handside of Equation (22) in arXiv:2210.13339
+      double var_target = ds[i] * ds[i] / k;
+      if (weights) {
+        var_target -= ds[i] * ds[i] / d;
+        for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+          var_target += A[j] * A[j];
+      }
+      FloatType c = cs[i];
+      // stands for left handside of Equation (22) in arXiv:2210.13339
```

Comment:
Same.

Reply:
This loop looks like it's doing fixed-point iteration to solve c, so I think the comment is incorrect.

Reply:
For both the cuda and cpu implementations, the O(nlogn) time computation described in the paper was slower. This implementation has complexity O(iters * n) time where iters is the number of iterations to converge. The larger n, usually the smaller number of iterations required. For most cases where fanout << n, it converges in a single iteration.

Reply:
Yes.  I'm just talking about this comment:
```
// stands for left handside of Equation (22) in arXiv:2210.13339
```
where it should be more like
```
// Compute c_s in Equation (22) via fixed-point iteration
```

Reply:
var_1 really stands for the lefthandside though.

Reply:
Though I could add an extra comment saying that fixed point iterations are utilized to compute c_s.